certbot/ocsp.py

```diff
@@ -65,33 +87,103 @@ def ocsp_revoked(self, cert_path, chain_path):
+        # TODO: (adferrand 2019-11-01) Following line can be replaced with a direct call to
+        #  response_ocsp.signature_hash_algorithm once cryptography 2.5 is released.
+        #  See https://github.com/pyca/cryptography/issues/4680
+        chosen_hash = x509._SIG_OIDS_TO_HASH[response_ocsp.signature_algorithm_oid]  # pylint: disable=protected-access
+    except KeyError:
+        raise UnsupportedAlgorithm(
+            "Signature algorithm OID:{0} not recognized"
+            .format(response_ocsp.signature_algorithm_oid)
+        )
+
+    issuer_cert.public_key().verify(
+        response_ocsp.signature,
+        response_ocsp.tbs_response_bytes,
+        padding.PSS(
+            mgf=padding.MGF1(chosen_hash),
+            salt_length=padding.PSS.MAX_LENGTH
+        ),
+        chosen_hash
+    )
+
+
+def _determine_ocsp_server(cert_path):
+    # type: (str) -> Tuple[Optional[str], Optional[str]]
+    """Extract the OCSP server host from a certificate.
+
+    :param str cert_path: Path to the cert we're checking OCSP for
+    :rtype tuple:
+    :returns: (OCSP server URL or None, OCSP server host or None)
+
+    """
+    with open(cert_path, 'rb') as file_handler:
+        cert = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
+    try:
+        extension = cert.extensions.get_extension_for_class(x509.AuthorityInformationAccess)
+        ocsp_oid = x509.AuthorityInformationAccessOID.OCSP
+        descriptions = [description for description in extension.value
+                        if description.access_method == ocsp_oid]
+
+        url = descriptions[0].access_location.value
+    except (x509.ExtensionNotFound, IndexError):
+        logger.info("Cannot extract OCSP URI from %s", cert_path)
+        return None, None
+
+    url = url.rstrip()
+    host = url.partition("://")[2].rstrip("/")
+
+    if host:
+        return url, host
+    else:
+        logger.info("Cannot process OCSP host from URL (%s) in cert at %s", url, cert_path)
+        return None, None
+
+
+def _check_ocsp_cryptography(cert_path, chain_path, url):
+    # type: (str, str, str) -> bool
+    # Retrieve OCSP response
+    with open(chain_path, 'rb') as file_handler:
+        issuer = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
+    with open(cert_path, 'rb') as file_handler:
+        cert = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
+    builder = ocsp.OCSPRequestBuilder()
+    builder = builder.add_certificate(cert, issuer, hashes.SHA1())
+    request = builder.build()
+    request_binary = request.public_bytes(serialization.Encoding.DER)
+    response = requests.post(url, data=request_binary,
+                             headers={'Content-Type': 'application/ocsp-request'})
+    if response.status_code != 200:
+        logger.info("OCSP check failed for %s (are we offline?)", cert_path)
+        return False
+    response_ocsp = ocsp.load_der_ocsp_response(response.content)
```

Comment:
I think we may need to change some things around after getting this response.

1. Looking at [RFC 6960](https://tools.ietf.org/html/rfc6960#section-3.2), we should probably make some other checks here.
2. We fail in an ugly way if the OCSP responder returns a failure response. The way I did this was try using a hash other than `SHA1()` above. When I did this with `--debug --verbose` on the command line, I got:
```
Exiting abnormally:
Traceback (most recent call last):
  File "/usr/local/bin/certbot", line 11, in <module>
    load_entry_point('certbot', 'console_scripts', 'certbot')()
  File "/home/bmw/src/certbot/certbot/main.py", line 1364, in main
    return config.func(config, plugins)
  File "/home/bmw/src/certbot/certbot/main.py", line 1038, in certificates
    cert_manager.certificates(config)
  File "/home/bmw/src/certbot/certbot/cert_manager.py", line 89, in certificates
    _describe_certs(config, parsed_certs, parse_failures)
  File "/home/bmw/src/certbot/certbot/cert_manager.py", line 355, in _describe_certs
    notify(_report_human_readable(config, parsed_certs))
  File "/home/bmw/src/certbot/certbot/cert_manager.py", line 340, in _report_human_readable
    certinfo.append(human_readable_cert_info(config, cert))
  File "/home/bmw/src/certbot/certbot/cert_manager.py", line 266, in human_readable_cert_info
    if checker.ocsp_revoked(cert.cert, cert.chain):
  File "/home/bmw/src/certbot/certbot/ocsp.py", line 69, in ocsp_revoked
    return _check_ocsp_cryptography(cert_path, chain_path, url)
  File "/home/bmw/src/certbot/certbot/ocsp.py", line 169, in _check_ocsp_cryptography
    _check_ocsp_response_signature(response_ocsp, issuer)
  File "/home/bmw/src/certbot/certbot/ocsp.py", line 99, in _check_ocsp_response_signature
    chosen_hash = x509._SIG_OIDS_TO_HASH[response_ocsp.signature_algorithm_oid]  # pylint: disable=protected-access
  File "/home/bmw/src/certbot/venv/local/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/ocsp.py", line 30, in wrapper
    "OCSP response status is not successful so the property "
ValueError: OCSP response status is not successful so the property has no value
```

Reply:
I resolved the issue 2., I am looking on the RFC to implement further checks.

Reply:
Ok, it added several checks from RFC 6960:
* certificate corresponds,
* issuer corresponds,
* thisUpdate/nextUpdate correspond

Except for checking responder is authorized, all optional checks from section 3.2 are covered.